File: app/src/main/java/com/scheffer/erik/bakingapp/adapters/RecipeAdapter.java
package com.scheffer.erik.bakingapp.adapters;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.scheffer.erik.bakingapp.R;
import com.scheffer.erik.bakingapp.StepListActivity;
import com.scheffer.erik.bakingapp.models.Ingredient;
import com.scheffer.erik.bakingapp.models.Recipe;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

import static com.scheffer.erik.bakingapp.StepListActivity.RECIPE_EXTRA_KEY;

public class RecipeAdapter extends RecyclerView.Adapter<RecipeAdapter.ViewHolder> {

    public static final String SHARED_PREFERENCES_NAME = "sharedpref";
    public static final String INGREDIENTS_PREFERENCES_KEY = "ingredients-pref-key";

    private List<Recipe> recipes;

    public RecipeAdapter(List<Recipe> recipes) {
        this.recipes = recipes;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                                  .inflate(R.layout.recipe_item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        Recipe recipe = recipes.get(position);
        holder.recipeNameText.setText(recipe.getName());
        holder.recipe = recipe;
    }

    @Override
    public int getItemCount() {
        return recipes.size();
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.recipe_name_text)
        TextView recipeNameText;

        Recipe recipe;

        ViewHolder(final View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);

            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    saveIngredientsOnSharedPreferences(itemView.getContext(), recipe);

                    Intent intent = new Intent(itemView.getContext(), StepListActivity.class);
                    intent.putExtra(RECIPE_EXTRA_KEY, recipe);
                    itemView.getContext().startActivity(intent);
                }
            });
        }

        private void saveIngredientsOnSharedPreferences(Context context, Recipe recipe) {
            SharedPreferences sharedPref = context.getSharedPreferences(
                    SHARED_PREFERENCES_NAME,
                    Context.MODE_PRIVATE);
            SharedPreferences.Editor editor = sharedPref.edit();
            StringBuilder builder = new StringBuilder();
            for (Ingredient ingredient : recipe.getIngredients()) {
                builder.append(context.getResources()
                                      .getString(R.string.ingredient_description,
                                                 ingredient.getQuantity(),
                                                 ingredient.getMeasure(),
                                                 ingredient.getIngredient())).append(";");
            }
            editor.putString(INGREDIENTS_PREFERENCES_KEY, builder.toString());
            editor.apply();
        }
    }
}

File: app/src/main/java/com/scheffer/erik/bakingapp/recyclerviewadapters/RecipeAdapter.kt
package com.scheffer.erik.bakingapp.recyclerviewadapters

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.scheffer.erik.bakingapp.R
import com.scheffer.erik.bakingapp.activities.StepListActivity
import com.scheffer.erik.bakingapp.activities.StepListActivity.Companion.RECIPE_EXTRA_KEY
import com.scheffer.erik.bakingapp.models.Recipe
import com.squareup.picasso.Picasso
import org.jetbrains.anko.startActivity

class RecipeAdapter(private val recipes: List<Recipe>,
                    private val context: Context) : RecyclerView.Adapter<RecipeAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
            ViewHolder(LayoutInflater.from(parent.context)
                               .inflate(R.layout.recipe_item, parent, false))

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val recipe = recipes[position]
        holder.recipeNameText.text = recipe.name
        holder.recipe = recipe

        if (!recipe.image.isEmpty()) {
            holder.recipeImage.visibility = View.VISIBLE
            Picasso.with(context)
                    .load(recipe.image)
                    .into(holder.recipeImage)
        } else {
            holder.recipeImage.visibility = View.GONE
        }
    }

    override fun getItemCount() = recipes.size

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val recipeNameText: TextView = itemView.findViewById(R.id.recipe_name_text)
        val recipeImage: ImageView = itemView.findViewById(R.id.recipe_image)
        lateinit var recipe: Recipe

        init {
            itemView.setOnClickListener {
                saveIngredientsOnSharedPreferences(itemView.context, recipe)
                itemView.context.startActivity<StepListActivity>(RECIPE_EXTRA_KEY to recipe)
            }
        }

        private fun saveIngredientsOnSharedPreferences(context: Context, recipe: Recipe) =
                context.getSharedPreferences(
                        SHARED_PREFERENCES_NAME,
                        Context.MODE_PRIVATE).edit().apply {
                    val builder = StringBuilder()
                    for (ingredient in recipe.ingredients) {
                        builder.append(context.resources
                                               .getString(R.string.ingredient_description,
                                                          ingredient.quantity,
                                                          ingredient.measure,
                                                          ingredient.ingredient)).append(";")
                    }
                    putString(INGREDIENTS_PREFERENCES_KEY, builder.toString())
                }.apply()
    }

    companion object {
        const val SHARED_PREFERENCES_NAME = "sharedpref"
        const val INGREDIENTS_PREFERENCES_KEY = "ingredients-pref-key"
    }
}
